File: ts/project/.sdk/tm/scala/utility/Fetcher.scala
package SCALAPACKAGE.utility

import java.net.{InetSocketAddress, ProxySelector, URI}
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.util.{LinkedHashMap, List => JList, Map => JMap}
import java.util.function.{BiFunction, Supplier}
import SCALAPACKAGE.core._
import SCALAPACKAGE.utility.struct.Struct

object Fetcher {

  def defaultHttpFetch(fullurl: String, fetchdef: JMap[String, Object]): JMap[String, Object] = {
    var method = fetchdef.get("method") match { case s: String => s; case _ => "" }
    if ("" == method) method = "GET"

    var bodyPublisher = HttpRequest.BodyPublishers.noBody()
    fetchdef.get("body") match {
      case b: String if b != "" => bodyPublisher = HttpRequest.BodyPublishers.ofString(b)
      case _ =>
    }

    val reqb =
      try HttpRequest.newBuilder(URI.create(fullurl)).method(method, bodyPublisher)
      catch { case e: RuntimeException => throw new RuntimeException("fetch: invalid url: " + fullurl, e) }

    var hasUA = false
    fetchdef.get("headers") match {
      case hm: JMap[_, _] =>
        val it = hm.asInstanceOf[JMap[String, Object]].entrySet().iterator()
        while (it.hasNext) {
          val e = it.next()
          e.getValue match {
            case v: String =>
              if ("user-agent".equalsIgnoreCase(e.getKey)) hasUA = true
              reqb.setHeader(e.getKey, v)
            case _ =>
          }
        }
      case _ =>
    }
    if (!hasUA) reqb.setHeader("User-Agent", "Mozilla/5.0 (compatible; ProjectNameSDK/1.0)")

    val clientb = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL)
    fetchdef.get("proxy") match {
      case proxy: String if proxy != "" =>
        try {
          val proxyUri = URI.create(proxy)
          val port = if (proxyUri.getPort < 0) 80 else proxyUri.getPort
          clientb.proxy(ProxySelector.of(new InetSocketAddress(proxyUri.getHost, port)))
        } catch { case _: RuntimeException => }
      case _ =>
    }

    val resp =
      try clientb.build().send(reqb.build(), HttpResponse.BodyHandlers.ofString())
      catch { case e: Exception => throw new RuntimeException("fetch: " + e.getMessage, e) }

    val headers = new LinkedHashMap[String, Object]()
    val hit = resp.headers().map().entrySet().iterator()
    while (hit.hasNext) {
      val h = hit.next()
      val vals = h.getValue
      if (vals.size() == 1) headers.put(h.getKey.toLowerCase, vals.get(0))
      else headers.put(h.getKey.toLowerCase, String.join(", ", vals))
    }

    val bodyText = if (resp.body() == null) "" else resp.body()
    val jsonBody: Object = if (bodyText.isEmpty) null else Json.parseOrNull(bodyText)
    val jsonSupplier: Supplier[Object] = () => jsonBody

    val out = new LinkedHashMap[String, Object]()
    out.put("status", java.lang.Integer.valueOf(resp.statusCode()))
    out.put("statusText", statusText(resp.statusCode()))
    out.put("headers", headers)
    out.put("json", jsonSupplier)
    out.put("body", bodyText)
    out
  }

  def statusText(status: Int): String = status match {
    case 200 => "OK"
    case 201 => "Created"
    case 202 => "Accepted"
    case 204 => "No Content"
    case 301 => "Moved Permanently"
    case 302 => "Found"
    case 304 => "Not Modified"
    case 400 => "Bad Request"
    case 401 => "Unauthorized"
    case 403 => "Forbidden"
    case 404 => "Not Found"
    case 405 => "Method Not Allowed"
    case 408 => "Request Timeout"
    case 409 => "Conflict"
    case 422 => "Unprocessable Entity"
    case 425 => "Too Early"
    case 429 => "Too Many Requests"
    case 500 => "Internal Server Error"
    case 502 => "Bad Gateway"
    case 503 => "Service Unavailable"
    case 504 => "Gateway Timeout"
    case _ => ""
  }

  def fetcher(ctx: Context, fullurl: String, fetchdef: JMap[String, Object]): Object = {
    if ("live" != ctx.client.mode) {
      throw ctx.makeError("fetch_mode_block",
        "Request blocked by mode: \"" + ctx.client.mode + "\" (URL was: \"" + fullurl + "\")")
    }

    val options = ctx.client.optionsMap()
    if (java.lang.Boolean.TRUE == Struct.getpath(options, java.util.List.of("feature", "test", "active"))) {
      throw ctx.makeError("fetch_test_block",
        "Request blocked as test feature is active (URL was: \"" + fullurl + "\")")
    }

    var sysFetch = Struct.getpath(options, java.util.List.of("system", "fetch"))
    if (sysFetch eq Struct.UNDEF) sysFetch = null

    if (sysFetch == null) return defaultHttpFetch(fullurl, fetchdef)

    sysFetch match {
      case bf: BiFunction[_, _, _] =>
        bf.asInstanceOf[BiFunction[String, JMap[String, Object], Object]].apply(fullurl, fetchdef)
      case _ =>
        throw ctx.makeError("fetch_invalid", "system.fetch is not a valid function")
    }
  }
}

File: ts/project/.sdk/tm/rb/utility/fetcher.rb
# ProjectName SDK utility: fetcher
require_relative 'struct/voxgig_struct'
require 'net/http'
require 'uri'
require 'json'

module ProjectNameUtilities
  # Started Net::HTTP connections keyed by scheme, host, port and proxy,
  # checked out for one request and returned. A Net::HTTP serves one
  # request at a time, so the pool holds idle ones rather than sharing one.
  HTTP_POOL = Hash.new { |h, k| h[k] = [] }
  HTTP_POOL_LOCK = Mutex.new
  HTTP_POOL_IDLE_MAX = 4

  def self.http_checkout(key, uri, proxy)
    http = HTTP_POOL_LOCK.synchronize { HTTP_POOL[key].pop }
    return http if http && http.started?

    if proxy
      purl = URI.parse(proxy)
      http = Net::HTTP.new(uri.host, uri.port, purl.host, purl.port, purl.user, purl.password)
    else
      http = Net::HTTP.new(uri.host, uri.port)
    end
    http.use_ssl = (uri.scheme == "https")
    http.start
  end

  def self.http_checkin(key, http)
    kept = HTTP_POOL_LOCK.synchronize do
      HTTP_POOL[key].size < HTTP_POOL_IDLE_MAX && HTTP_POOL[key].push(http)
    end
    http.finish unless kept
  rescue IOError
    nil
  end

  def self.http_discard(http)
    http.finish if http.started?
  rescue IOError
    nil
  end

  DefaultHttpFetch = ->(fullurl, fetchdef) {
    method_str = fetchdef["method"] || "GET"
    body_str = fetchdef["body"]
    headers = fetchdef["headers"] || {}

    begin
      uri = URI.parse(fullurl)
      proxy = fetchdef["proxy"]
      proxy = nil unless proxy.is_a?(String) && !proxy.empty?
      key = [uri.scheme, uri.host, uri.port, proxy]

      klass = case method_str.upcase
              when "POST" then Net::HTTP::Post
              when "PUT" then Net::HTTP::Put
              when "DELETE" then Net::HTTP::Delete
              when "PATCH" then Net::HTTP::Patch
              else Net::HTTP::Get
              end

      request = klass.new(uri)
      has_ua = false
      headers.each do |k, v|
        next unless v.is_a?(String)
        has_ua = true if k.to_s.downcase == 'user-agent'
        request[k] = v.to_s
      end
      # Default User-Agent — Net::HTTP sets "Ruby" which some CDNs block.
      # Use a Mozilla-shaped UA unless the caller already set one.
      request['User-Agent'] = 'Mozilla/5.0 (compatible; ProjectNameSDK/1.0)' unless has_ua
      request.body = body_str if body_str.is_a?(String)

      http = ProjectNameUtilities.http_checkout(key, uri, proxy)
      begin
        # Net::HTTP retries idempotent requests; an EOF can follow a committed POST.
        resp = http.request(request)
      rescue StandardError
        ProjectNameUtilities.http_discard(http)
        raise
      end
      ProjectNameUtilities.http_checkin(key, http)

      resp_headers = {}
      resp.each_header { |k, v| resp_headers[k.downcase] = v }

      json_body = nil
      begin
        json_body = JSON.parse(resp.body) if resp.body && !resp.body.empty?
      rescue JSON::ParserError
      end

      return {
        "status" => resp.code.to_i,
        "statusText" => resp.message,
        "headers" => resp_headers,
        "json" => -> { json_body },
        "body" => resp.body,
      }, nil
    rescue StandardError => e
      # Network-level failures (DNS, TCP, TLS, timeouts) — return a synthesized
      # response with status 0 so callers can branch on result.ok like any
      # other failed request, instead of seeing an unhandled exception.
      return {
        "status" => 0,
        "statusText" => "#{e.class}: #{e.message}",
        "headers" => {},
        "json" => -> { nil },
        "body" => nil,
      }, nil
    end
  }

  Fetcher = ->(ctx, fullurl, fetchdef) {
    if ctx.client.mode != "live"
      return nil, ctx.make_error("fetch_mode_block",
        "Request blocked by mode: \"#{ctx.client.mode}\" (URL was: \"#{fullurl}\")")
    end

    options = ctx.client.options_map
    if VoxgigStruct.getpath(options, "feature.test.active") == true
      return nil, ctx.make_error("fetch_test_block",
        "Request blocked as test feature is active (URL was: \"#{fullurl}\")")
    end

    sys_fetch = VoxgigStruct.getpath(options, "system.fetch")

    return DefaultHttpFetch.call(fullurl, fetchdef) if sys_fetch.nil?
    return sys_fetch.call(fullurl, fetchdef) if sys_fetch.is_a?(Proc)

    return nil, ctx.make_error("fetch_invalid", "system.fetch is not a valid function")
  }
end
